cmake/ChooseBlas.cmake

```diff
@@ -16,43 +16,199 @@
+  set(mxnet_LINKER_LIBS ${mxnet_LINKER_LIBS} ${MKLML_LIBRARIES} PARENT_SCOPE)
+
+  set(MKL_FOUND ${MKL_FOUND} PARENT_SCOPE)
+  set(MKLROOT ${MKLROOT} PARENT_SCOPE)
+
+  set(BLAS MKL PARENT_SCOPE)
+
+endfunction()
+
+function(try_accelerate)
+  if(NOT APPLE)
+    return()
+  endif()
+
+  if(BLAS MATCHES "[Mm][Kk][Ll]")
+    return()
+  endif()
+
+  if(${USE_ACCELERATE_IF_AVAILABLE})
+    message(STATUS "Trying to enable accelerate framework due to USE_ACCELERATE_IF_AVAILABLE")
+    find_package(Accelerate)
+    if(${Accelerate_FOUND})
+      message(STATUS "Accelerate framework found")
+      set(BLAS Accelerate PARENT_SCOPE)
     else()
-      set(BLAS "MKL")
+      message(STATUS "Accelerate framework not found")
     endif()
   endif()
-endif()
+endfunction()
 
-if(BLAS STREQUAL "Atlas" OR BLAS STREQUAL "atlas")
+try_mkl()
+try_mklml()
+try_mkldnn()
+try_accelerate()
+
+# cmake regexp does not support case insensitive match (?i) or //i
+if(BLAS MATCHES "[Aa][Tt][Ll][Aa][Ss]")
+  message(STATUS "Using Atlas for BLAS")
+
+  set(Atlas_NEED_LAPACK ${USE_LAPACK})
   find_package(Atlas REQUIRED)
-  include_directories(SYSTEM ${Atlas_INCLUDE_DIR})
-  list(APPEND mshadow_LINKER_LIBS ${Atlas_LIBRARIES})
+
+  include_directories(SYSTEM ${Atlas_INCLUDE_DIRS})
+  set(mxnet_LINKER_LIBS ${mxnet_LINKER_LIBS} ${Atlas_LIBRARIES})
+
   add_definitions(-DMSHADOW_USE_CBLAS=1)
   add_definitions(-DMSHADOW_USE_MKL=0)
-elseif(BLAS STREQUAL "Open" OR BLAS STREQUAL "open")
+
+  switch_lapack(${USE_LAPACK} AND ${Atlas_LAPACK_FOUND})
+
+  return()
+endif()
+
+if(BLAS MATCHES "[Oo][Pp][Ee][Nn]")
+  message(STATUS "Using OpenBLAS for BLAS")
+
+  set(OpenBLAS_NEED_LAPACK ${USE_LAPACK})
+
   find_package(OpenBLAS REQUIRED)
-  include_directories(SYSTEM ${OpenBLAS_INCLUDE_DIR})
-  list(APPEND mshadow_LINKER_LIBS ${OpenBLAS_LIB})
+
+  include_directories(SYSTEM ${OpenBLAS_INCLUDE_DIRS})
+  set(mxnet_LINKER_LIBS ${mxnet_LINKER_LIBS} ${OpenBLAS_LIBRARIES})
+
   add_definitions(-DMSHADOW_USE_CBLAS=1)
   add_definitions(-DMSHADOW_USE_MKL=0)
-elseif(BLAS STREQUAL "MKL" OR BLAS STREQUAL "mkl")
+
+  switch_lapack(${USE_LAPACK} AND ${OpenBLAS_LAPACK_FOUND})
+
+  return()
+endif()
+
+if(BLAS MATCHES "[Mm][Kk][Ll]")
+  message(STATUS "Using MKL for BLAS")
+
   find_package(MKL REQUIRED)
+
   include_directories(SYSTEM ${MKL_INCLUDE_DIR})
-  list(APPEND mshadow_LINKER_LIBS ${MKL_LIBRARIES})
+  set(mxnet_LINKER_LIBS ${mxnet_LINKER_LIBS} ${MKL_LIBRARIES})
+
   add_definitions(-DMSHADOW_USE_CBLAS=0)
   add_definitions(-DMSHADOW_USE_MKL=1)
-elseif(BLAS STREQUAL "apple")
+
+  if(USE_LAPACK)
+    include(CheckFunctionExists)
+    set(CMAKE_REQUIRED_LIBRARIES ${MKL_LIBRARIES})
+    check_function_exists("cgees_" LAPACK_FOUND)
+
+    switch_lapack(${LAPACK_FOUND} OR False)
+
+    return()
+  endif()
+endif()
+
+if(BLAS MATCHES "([Aa][Pp][Pp][Ll][Ee]|[Aa][Cc][Cc][Ee][Ll][Ee][Rr][Aa][Tt][Ee])")
+  if(NOT APPLE)
+    message(FATAL_ERROR "Apple BLAS framework is available only on MAC")
```

Comment:
Apple BLAS == Apple Accelerate right? Can we keep the names consistent? So...
"Apple Accelerate framework's BLAS feature is available only on macOS"